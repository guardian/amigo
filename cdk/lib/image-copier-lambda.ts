import type { GuStackProps } from "@guardian/cdk/lib/constructs/core";
import { GuStack } from "@guardian/cdk/lib/constructs/core";
import type { App } from "aws-cdk-lib";
import { CfnParameter, Duration, Fn } from "aws-cdk-lib";
import { Effect, PolicyStatement } from "aws-cdk-lib/aws-iam";
import { Code, Function, Runtime } from "aws-cdk-lib/aws-lambda";
import { SnsEventSource } from "aws-cdk-lib/aws-lambda-event-sources";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { Topic } from "aws-cdk-lib/aws-sns";

export class ImageCopierLambda extends GuStack {
  constructor(scope: App, id: string, props: GuStackProps) {
    super(scope, id, props);

    const functionCodeBucket = Bucket.fromBucketName(this, "function-code-bucket", "deploy-tools-dist");

    const kmsKeyArnParam = new CfnParameter(this, "KmsKeyArn", {
      description: "Override the default KMS key if required",
      type: "String",
      default: "",
    });

    const kmsKeyArn = Fn.importValue("amigo-imagecopier-key");

    const housekeepingTopicParam = new CfnParameter(this, "AmigoHousekeepingTopicArn", {
      description: "The housekeeping SNS topic to subscribe to",
      type: "String",
    });

    const copierTopicParam = new CfnParameter(this, "AmigoTopicArn", {
      description: "The SNS topic to subscribe to",
      type: "String",
    });

    const loggingPolicy = new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ["logs:CreateLogGroup", "logs:CreateLogStream", "logs:PutLogEvents"],
      resources: ["arn:aws:logs:*:*:*"],
    });

    const copierLambda = new Function(this, "ImageCopierLambda", {
      description: "Lambda for copying and encrypting AMIgo baked AMIs",
      runtime: Runtime.JAVA_11,
      memorySize: 512,
      handler: "com.gu.imageCopier.LambdaEntrypoint::run",
      timeout: Duration.seconds(30),
      environment: {
        ACCOUNT_ID: this.account,
        KMS_KEY_ARN: kmsKeyArn,
        ENCRYPTED_TAG_VALUE: "true",
      },
      code: Code.fromBucket(functionCodeBucket, `${this.stack}/${this.stage}/imagecopier/imagecopier.zip`),
      initialPolicy: [
        loggingPolicy,
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ["ec2:CopyImage", "ec2:CreateTags"],
          resources: ["*"],
        }),
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ["kms:Encrypt", "kms:Decrypt", "kms:CreateGrant", "kms:GenerateDataKey*", "kms:DescribeKey"],
          resources: [Fn.importValue("amigo-imagecopier-key")],
        }),
      ],
    });

    this.overrideLogicalId(copierLambda, {
      logicalId: "ImageCopierLambda",
      reason: "To gain confidence during the migration to CDK",
    });

    const copierTopic = Topic.fromTopicArn(
      this,
      `ImageCopierLambda-SnsExistingIncomingEventsTopic`,
      copierTopicParam.valueAsString
    );

    copierLambda.addEventSource(new SnsEventSource(copierTopic));

    const housekeepingLambda = new Function(this, "HousekeepingLambda", {
      description: "Lambda for housekeeping AMIgo baked AMIs in other accounts",
      runtime: Runtime.JAVA_11,
      memorySize: 512,
      handler: "com.gu.imageCopier.LambdaEntrypoint::housekeeping",
      timeout: Duration.seconds(30),
      environment: {
        ACCOUNT_ID: this.account,
        KMS_KEY_ARN: kmsKeyArn,
        ENCRYPTED_TAG_VALUE: "true",
      },
      code: Code.fromBucket(functionCodeBucket, `${this.stack}/${this.stage}/imagecopier/imagecopier.zip`),
      initialPolicy: [
        loggingPolicy,
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ["ec2:DescribeImages", "ec2:DeregisterImage", "ec2:DeleteSnapshot"],
          resources: ["*"],
        }),
      ],
    });

    this.overrideLogicalId(housekeepingLambda, {
      logicalId: "HousekeepingLambda",
      reason: "To gain confidence during the migration to CDK",
    });

    const housekeepingTopic = Topic.fromTopicArn(
      this,
      `HousekeepingLambda-SnsExistingIncomingEventsTopic`,
      housekeepingTopicParam.valueAsString
    );

    housekeepingLambda.addEventSource(new SnsEventSource(housekeepingTopic));
  }
}
