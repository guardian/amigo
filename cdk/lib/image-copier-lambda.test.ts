import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { ImageCopierLambda } from "./image-copier-lambda";

describe("the lambda stack", () => {
  it("matches the snapshot", () => {
    const app = new App();
    const stack = new ImageCopierLambda(app, "IntegrationTest", {
      stack: "cdk",
      stage: "TEST",
      version: "v4",
    });
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
