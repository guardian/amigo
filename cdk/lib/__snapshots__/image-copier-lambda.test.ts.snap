// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the lambda stack matches the snapshot 1`] = `
Object {
  "Metadata": Object {
    "gu:cdk:constructs": Array [],
    "gu:cdk:version": "TEST",
  },
  "Parameters": Object {
    "AmigoHousekeepingTopicArn": Object {
      "Description": "The housekeeping SNS topic to subscribe to",
      "Type": "String",
    },
    "AmigoTopicArn": Object {
      "Description": "The SNS topic to subscribe to",
      "Type": "String",
    },
    "KmsKeyArn": Object {
      "Default": "",
      "Description": "Override the default KMS key if required",
      "Type": "String",
    },
  },
  "Resources": Object {
    "HousekeepingLambda": Object {
      "DependsOn": Array [
        "HousekeepingLambdaServiceRoleDefaultPolicy003C9832",
        "HousekeepingLambdaServiceRole70148A54",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "deploy-tools-dist",
          "S3Key": "cdk/TEST/imagecopier/image-copier.zip",
        },
        "Description": "Lambda for housekeeping AMIgo baked AMIs in other accounts",
        "Environment": Object {
          "Variables": Object {
            "ACCOUNT_ID": Object {
              "Ref": "AWS::AccountId",
            },
            "ENCRYPTED_TAG_VALUE": "true",
            "KMS_KEY_ARN": Object {
              "Fn::ImportValue": "amigo-imagecopier-key",
            },
          },
        },
        "Handler": "com.gu.imageCopier.LambdaEntrypoint::housekeeping",
        "MemorySize": 512,
        "Role": Object {
          "Fn::GetAtt": Array [
            "HousekeepingLambdaServiceRole70148A54",
            "Arn",
          ],
        },
        "Runtime": "java11",
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/amigo",
          },
          Object {
            "Key": "Stack",
            "Value": "cdk",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "HousekeepingLambdaAllowInvokeIntegrationTestHousekeepingLambdaSnsExistingIncomingEventsTopic1B0CC4879A69C8A5": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "HousekeepingLambda",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": Object {
          "Ref": "AmigoHousekeepingTopicArn",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "HousekeepingLambdaHousekeepingLambdaSnsExistingIncomingEventsTopicDFF8A27F": Object {
      "Properties": Object {
        "Endpoint": Object {
          "Fn::GetAtt": Array [
            "HousekeepingLambda",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "Region": Object {
          "Fn::Select": Array [
            3,
            Object {
              "Fn::Split": Array [
                ":",
                Object {
                  "Ref": "AmigoHousekeepingTopicArn",
                },
              ],
            },
          ],
        },
        "TopicArn": Object {
          "Ref": "AmigoHousekeepingTopicArn",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "HousekeepingLambdaServiceRole70148A54": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/amigo",
          },
          Object {
            "Key": "Stack",
            "Value": "cdk",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "HousekeepingLambdaServiceRoleDefaultPolicy003C9832": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:logs:*:*:*",
            },
            Object {
              "Action": Array [
                "ec2:DescribeImages",
                "ec2:DeregisterImage",
                "ec2:DeleteSnapshot",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "HousekeepingLambdaServiceRoleDefaultPolicy003C9832",
        "Roles": Array [
          Object {
            "Ref": "HousekeepingLambdaServiceRole70148A54",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ImageCopierLambda": Object {
      "DependsOn": Array [
        "ImageCopierLambdaServiceRoleDefaultPolicy82F190E9",
        "ImageCopierLambdaServiceRole77B845A3",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "deploy-tools-dist",
          "S3Key": "cdk/TEST/imagecopier/image-copier.zip",
        },
        "Description": "Lambda for copying and encrypting AMIgo baked AMIs",
        "Environment": Object {
          "Variables": Object {
            "ACCOUNT_ID": Object {
              "Ref": "AWS::AccountId",
            },
            "ENCRYPTED_TAG_VALUE": "true",
            "KMS_KEY_ARN": Object {
              "Fn::ImportValue": "amigo-imagecopier-key",
            },
          },
        },
        "Handler": "com.gu.imageCopier.LambdaEntrypoint::run",
        "MemorySize": 512,
        "Role": Object {
          "Fn::GetAtt": Array [
            "ImageCopierLambdaServiceRole77B845A3",
            "Arn",
          ],
        },
        "Runtime": "java11",
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/amigo",
          },
          Object {
            "Key": "Stack",
            "Value": "cdk",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ImageCopierLambdaAllowInvokeIntegrationTestImageCopierLambdaSnsExistingIncomingEventsTopicD22ACB92E3D10371": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "ImageCopierLambda",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": Object {
          "Ref": "AmigoTopicArn",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ImageCopierLambdaImageCopierLambdaSnsExistingIncomingEventsTopicB31F3C6C": Object {
      "Properties": Object {
        "Endpoint": Object {
          "Fn::GetAtt": Array [
            "ImageCopierLambda",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "Region": Object {
          "Fn::Select": Array [
            3,
            Object {
              "Fn::Split": Array [
                ":",
                Object {
                  "Ref": "AmigoTopicArn",
                },
              ],
            },
          ],
        },
        "TopicArn": Object {
          "Ref": "AmigoTopicArn",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "ImageCopierLambdaServiceRole77B845A3": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/amigo",
          },
          Object {
            "Key": "Stack",
            "Value": "cdk",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ImageCopierLambdaServiceRoleDefaultPolicy82F190E9": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:logs:*:*:*",
            },
            Object {
              "Action": Array [
                "ec2:CopyImage",
                "ec2:CreateTags",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "kms:Encrypt",
                "kms:Decrypt",
                "kms:CreateGrant",
                "kms:GenerateDataKey*",
                "kms:DescribeKey",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::ImportValue": "amigo-imagecopier-key",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ImageCopierLambdaServiceRoleDefaultPolicy82F190E9",
        "Roles": Array [
          Object {
            "Ref": "ImageCopierLambdaServiceRole77B845A3",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
